File: BOJ/25_최단경로/11404_플로이드.cpp
#include <cstdio>
#include <iostream>
#include <vector>

#define MAX 100
#define INF 987654321
using namespace std;

int N,M;
int adjs[MAX+1][MAX+1];

void floyd() {
    for(int k=1; k<=N; k++){
        for(int i=1; i<=N; i++){
            for(int j=1; j<=N; j++){
                if(i == j) continue; 
                if(adjs[i][j] > adjs[i][k] + adjs[k][j])
                    adjs[i][j] = adjs[i][k] + adjs[k][j];
            }
        }
    }
}

int main() {
   
    int x,y,c;
    scanf("%d",&N);
    scanf("%d",&M);

    for(int i=1; i<=N; i++) {
        for(int j=1; j<=N; j++){
            adjs[i][j] = INF;
        }
    }

    for(int i=0; i<M; i++) {
        scanf("%d %d %d",&x,&y,&c);
        if(adjs[x][y] > c) adjs[x][y] = c;
    }
    floyd();

    for(int i=1; i<=N; i++) {
        for(int j=1; j<=N; j++) {
            if(adjs[i][j] == INF) printf("0 ");
            else printf("%d ",adjs[i][j]);
        }
        printf("\n");
    }



    return 0;
}

File: BOJ/25_최단경로/11404_플로이드.py
import sys

read = sys.stdin.readline
INF = int(1e9)

# 플로이드와샬 알고리즘은 출발-도착 > 출발 + 경유 + 도착 인 경우를 찾아주는 것
# 정점의 크기가 크면 할 수 없는 시간복잡도가 O(N^3) 인 알고리즘이다. 

def floyd() :
    for k in range(1,N+1) :
        for i in range(1,N+1) :
            for j in range(1,N+1) :
                if i == j : continue
                if adjs[i][j] > adjs[i][k] + adjs[k][j] :
                    adjs[i][j] = adjs[i][k] + adjs[k][j]



N = int(read())
M = int(read())
adjs = [[INF] * (N+1) for _ in range(N+1)]
for _ in range(M) :
    x,y,c = map(int,read().split())
    if adjs[x][y] > c :
        adjs[x][y] = c

floyd()

for i in range(1,N+1) :
    for j in range(1,N+1) :
        if adjs[i][j] == INF :
            print(0,end=' ')
        else :
            print(adjs[i][j],end=' ')
    print()